Celeste.Mod.mm/Mod/Registry/DecalRegistry.cs

```diff
@@ -193,47 +226,80 @@ public static class DecalRegistry {
+            List<string> localDecals = decalRegistry.Source.Map.Keys
+                .Where(s => s.StartsWith(basePath))
+                .Select(s => s.Substring(basePath.Length).TrimEnd(digits).ToLower())
+                .Distinct()
+                .ToList();
+
+            foreach (KeyValuePair<string, DecalInfo> decalRegistration in ReadDecalRegistryXml(decalRegistry)) {
+                string registeredPath = decalRegistration.Key;
+                DecalInfo info = decalRegistration.Value;
+
+                bool found = false;
+                if (registeredPath.EndsWith("*") || registeredPath.EndsWith("/")) {
+                    registeredPath = registeredPath.TrimEnd('*');
+                    foreach (string decalPath in localDecals) {
+                        // Wildcard matches must be longer than the subpath, and don't match on decals in subfolders
+                        if (decalPath.StartsWith(registeredPath) && decalPath.Length > registeredPath.Length 
+                            && decalPath.LastIndexOf('/') <= registeredPath.Length - 1) {
+                            RegisterDecal(decalPath, info);
+                            found = true;
+                        }
+                    }
+                } else if (localDecals.Contains(registeredPath)) {
+                    RegisterDecal(registeredPath, info);
+                    found = true;
+                } 
+                
+                if (!found) {
+                    Logger.Log(LogLevel.Warn, "Decal Registry", $"Could not find any decals in {decalRegistry.Source.Name} under path {decalRegistration.Key}");
+                }
             }
         }
 
         /// <summary>
-        /// Reads a DecalRegistry.xml file's contents
+        /// Reads a decal registry file and returns a sorted list of properties per decal path.
         /// </summary>
-        public static List<KeyValuePair<string, DecalInfo>> ReadDecalRegistryXml(string fileContents) {
-            // XmlElement file = Calc.LoadXML(path)["decals"];
+        private static List<KeyValuePair<string, DecalInfo>> ReadDecalRegistryXml(ModAsset decalRegistry) {
             XmlDocument doc = new XmlDocument();
-            doc.LoadXml(fileContents);
-            XmlElement file = doc["decals"];
+            doc.Load(decalRegistry.Stream);
 
             List<KeyValuePair<string, DecalInfo>> elements = new();
-            foreach (XmlNode node in file) {
+
+            foreach (XmlNode node in doc["decals"]) {
                 if (node is XmlElement decal) {
                     string decalPath = decal.Attr("path", null)?.ToLower();
                     if (decalPath == null) {
-                        Logger.Log(LogLevel.Warn, "Decal Registry", "Decal didn't have a path attribute!");
+                        Logger.Log(LogLevel.Warn, "Decal Registry", $"Decal in {decalRegistry.Source.Name} registry didn't have a path attribute!");
```

Comment:
While we're changing the error message, this sounds slightly strange. Maybe "is missing a path attribute" instead of "didn't have"?